user/sccsget.man
.\" $Id: sccsget.man,v 6.3 2023/01/20 09:27:57 tom Exp $
.de Es
.ne \\$1
.nr mE \\n(.f
.RS 5n
.sp .7
.nf
.nh
.ta 9n 17n 25n 33n 41n 49n
.ft CW
..
.de Eh
.ft \\n(mE
.fi
.hy \\n(HY
.RE
.sp .7
..
.\" Bulleted paragraph
.de bP
.ie n  .IP \(bu 4
.el    .IP \(bu 2
..
.TH SCCSGET 1
.SH NAME
.PP
\fBsccsget\fR \- SCCS check-out script
.SH USAGE
.PP
\fBsccsget\fI [options] [file-specifications]\fR
.SH SYNOPSIS
.PP
\fBSccsget\fR is a simple interface to \fIsccs\fR
(\fBs\fRource \fBc\fRode \fBc\fRontrol \fBs\fRystem),
which is used to extract files from \fIsccs\fR archives.
.SH DESCRIPTION
.PP
\fBSccsget\fR checks out (extracts) files from an \fIsccs\fR
archive which have been checked in using \fBsccsput\fR.
\fBSccsput\fR
checks in files, retaining their modification date as the check-in
date.
\fBSccsget\fR extracts a given version of one or more files,
and sets their modification date to match the check-in date.
.SH OPTIONS
.PP
\fBSccsget\fR is a directory-recursive interface
to the \fBgetdelta\fR utility (which invokes \fIsccs\fR's \fBget\fR)
which does special things with the file modification date.
Most of its options are passed to \fBget\fR, and then interpreted to determine
the correct modification date:
.TP
.BI \-c cutoff
specifies the \fIsccs\fR cutoff date, in the format
.Es
YY[MM[DD[HH[MM[SS]]]]]
.Eh
.RS
.PP
Punctuation in the date is permitted.
For example,
.Es
-c 77/2/2 9:22:25
-c 770202092225
.Eh
.PP
represent the same date.
.RE
.TP
.B \-f
forces \fBsccsget\fR to remove the current version
of the file before extracting the specified version.
Otherwise, \fBsccsget\fR will refuse to overwrite a file which already exists.
.TP
.B \-k
directs get to extract the specified file without
substituting sccs identification keywords.
You should use this option
if you intend to make updates to the file.
.TP
.B \-n
suppresses the \fBget\fR-operation.
Instead, \fBsccsget\fR echoes an "ok" message
for files which would be extracted.
.TP
.BI -r SID
specifies the \fIsccs\fR identification
string (SID) of the version of an \fIsccs\fR file to be retrieved.
.TP
.B \-s
suppresses most messages written by the \fBget\fR
utility (i.e., those written to standard output).
.SH OPERATIONS
.PP
To extract files using \fBsccsget\fR, you must provide one or more
names.
They may be either file or directory names.
.PP
\fBSccsget\fR resolves names until it finds one of the form "sccs/s.
\fIfile\fR" before invoking \fBget\fR.
If it finds a directory
(other than "SCCS"), \fBsccsget\fR will recur on itself to
extract files at lower levels.
.SH ENVIRONMENT
.PP
\fBSccsget\fR is a Bourne shell script.
On Apollo DOMAIN/IX, it
uses System\ 5 features, including \fIgetopt\ (1)\fR.
Environment
variables used by \fBsccsget\fR include:
.TP
.B SCCS_DIR
specifies the directory in which \fBgetdelta\fR will find the "s." files.
If not specified, \fBgetdelta\fR
assumes "SCCS".
.SH FILES
.PP
Sccsget uses the following files:
.TP
.B sccsget
the Bourne shell script
.TP
.B getdelta
a utility which invokes \fBget\fR,
and then modifies the file's modification date after check-out so
that the file's modification date matches the version-date selected
with the "-\fBc\fR" or "-\fBr\fR" option.
.SH ANTICIPATED CHANGES
.PP
None.
.SH SEE ALSO
.PP
getdelta, sccsput, ded, get(1)
.SH AUTHOR
.PP
Thomas Dickey (Software Productivity Consortium).


CHANGES
2023-01-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* bin/makefile.in:
	s/USER/AS_USER/ to avoid environment-variable conflict

	* package/sccs_tools-12.0.spec: amended the version-dependency

2023-01-20  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/getdelta/test/run_test.sh, src/sccs2rcs/test/run_test.sh,
	  src/sccs2rcs/test/do_test.sh:
	shellcheck warnings

	* src/putdelta/test/run_test.sh: fix shellcheck warnings

	* src/putdelta/test/run_test.sh:
	discard stderr (some variants complain profusely about missing keywords)

	* src/getdelta/test/run_test.sh:
	modify to ensure the files have SCCS keywords, since some variants do not handle this gracefully

	* user/sccs2rcs.man: check-manpage

	* user/fixsccs.man: RCS_BASE

	* user/sccsget.man, user/sccsput.man, user/getdelta.man,
	  user/putdelta.man:
	check-manpage

	* user/makefile.in: add manpage for fixsccs

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

	* package/debian/control: update standards

	* package/debian/compat: bump

2023-01-19  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/sccs_tools-12.0.spec:
	use td_lib package, rather than side-by-side configuration

	* configure: regen

	* aclocal.m4: provide for finding td_lib using pkg-config

	* package/debian/copyright, COPYING: bump

2022-12-31  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2022-10-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/makefile.in: use install-sh if needed

	* src/putdelta/test/run_test.sh: s/fgrep/grep/

	* package/debian/copyright: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* COPYING, package/debian/changelog, package/sccs_tools-12.0.spec: bump

2022-08-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2022-08-01

2021-03-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2021-01-10  Thomas E. Dickey  <dickey@invisible-island.net>

	* configure: regen

	* aclocal.m4: quoting fixes for shellcheck

	* src/getdelta/src/getdelta.c, src/getdelta/src/sccsget.c,
	  src/sccs2rcs/src/sccs2rcs.c:
	gcc-warnings

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* COPYING, package/debian/changelog, package/debian/copyright,
	  package/sccs_tools-12.0.spec:
	bump

2020-12-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

2020-12-07  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/getdelta/test/makefile.in, src/putdelta/test/makefile.in,
	  src/sccs2rcs/test/makefile.in, src/fixsccs/test/makefile.in,
	  src/getdelta/src/makefile.in, src/putdelta/src/makefile.in,
	  src/sccs2rcs/src/makefile.in, src/fixsccs/src/makefile.in:
	use VPATH

2020-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/debian/changelog,
	  package/sccs_tools-12.0.spec:
	bump

2020-11-30  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: update to 2020-11-14

2020-11-26  Thomas E. Dickey  <dickey@invisible-island.net>

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* configure.in: without-X

2020-08-16  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2020-08-17

2019-12-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/putdelta/src/sccsput.c: suppress pager if not a tty

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

	* src/putdelta/src/putdelta.c: add an fflush

2019-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/putdelta.man: mention the other options that putdelta handles

	* src/sccs2rcs/src/sccs2rcs.c: stricter gcc warnings

	* src/putdelta/src/putdelta.c, src/putdelta/src/sccsput.c:
	use DYN-argv lists to replace catarg()

	* package/debian/compat: bump

	* src/getdelta/src/getdelta.c: use DYN-argv lists to replace catarg()

	* user/sccsget.man, user/sccsput.man, user/putdelta.man,
	  user/getdelta.man:
	check-manpage fixes (rename Ex to Eh for consistency, other errata)

	* src/putdelta/test/run_test.sh: use sccs wrapper

	* src/getdelta/src/sccsget.c: use DYN-argv lists to replace catarg()

	* src/getdelta/test/run_test.sh: use sccs wrapper

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* package/debian/copyright, COPYING, package/debian/changelog,
	  package/sccs_tools-12.0.spec:
	bump

2019-09-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2019-09-10

2018-03-24  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/changelog: bump

	* package/sccs_tools-12.0.spec:
	bump, disable debug-packages, update ftp url

	* package/debian/watch: update ftp url

	* package/debian/rules: use Debian build-flags

2018-01-07  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/debian/changelog,
	  package/sccs_tools-12.0.spec:
	bump

	* configure: regen

	* aclocal.m4, configure.in: resync with my-autoconf

2017-11-25  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2017-11-07

	* config.sub: 2017-11-23

2015-07-04  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/makefile.in: improve clean-rule

	* package/sccs_tools-12.0.spec, package/debian/copyright,
	  package/debian/changelog, COPYING:
	bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2015-05-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2015-03-04

	* config.sub: 2015-03-08

2014-12-14  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/sccsput.man, user/putdelta.man, user/getdelta.man,
	  user/sccsget.man:
	adapt/reuse macros from cm_tools to improve groff's translation to html

2013-12-07  Thomas E. Dickey  <dickey@invisible-island.net>


src/getdelta/src/sccsget.c
/*
 * Title:	sccsget.c (sccs get-tree)
 * Author:	T.E.Dickey
 * Created:	23 May 1991 (from sccsget.sh and rcsget.c)
 * Modified:
 *		06 Dec 2019, use DYN-strings to replace catarg
 *		19 Jul 1994, added dir-only logic, made this work with the
 *			     SCCS_VAULT config, and defined repeated "-n"
 *			     option for debugging.  Use 'sccs_debug()'.
 *		23 Sep 1993, gcc warnings
 *		24 Oct 1991, converted to ANSI
 *		20 Jun 1991, pass-thru "-e" option, use 'shoarg()'
 *		
 * Function:	Use SCCS-get to checkout one or more files from the sccs-
 *		directory which is located in the current working directory.
 *		Then, use 'sccsdate' to set the modification date of the
 *		checked-out files according to the last delta date (rather than
 *		the current date, as SCCS assumes).
 *
 *		Options are designed to feed-thru to 'get(1)'.
 *
 * Environment:
 *		SCCS_DIR- name of sccs directory
 *
 * Options:	see 'usage()'
 */

#define	STR_PTYPES
#include	<ptypes.h>
#include	<dyn_str.h>
#include	<rcsdefs.h>
#include	<sccsdefs.h>

MODULE_ID("$Id: sccsget.c,v 6.13 2021/01/10 19:52:14 tom Exp $")

static ARGV *get_opts;
static const char *verb = "getdelta";
static int a_opt;		/* all-directory scan */
static int no_op;		/* no-op mode */
static int force;
static int debug;
static int dir_only;		/* expand names via archive-directories */
static const char *working_path;	/* used to resolve SCCS working/archive dirs */

static void
checkout(const char *name)
{
    ARGV *args;
    char *working = sccs2name(name, FALSE);

    if (!strncmp("p.", pathleaf(name), 2)) {
	return;
    }
    args = argv_init1(verb);
    argv_merge(&args, get_opts);
    if (no_op > 1)
	argv_append(&args, "-n");
    argv_append(&args, working);
    if (debug)
	show_argv(stdout, argv_values(args));
    if (no_op != 1) {
	if (executev(argv_values(args)) < 0)
	    failed(working);
    }
    argv_free(&args);
}

static void
SetWd(const char *path)
{
    if (debug)
	shoarg(stdout, "chdir", path);
    if (no_op != 1) {
	if (chdir(path) < 0)
	    failed(path);
    }
}

static int
fexists(const char *name)
{
    Stat_t sb;
    return (stat_file(name, &sb) >= 0);
}

/*ARGSUSED*/
static int
WALK_FUNC(do_archive)
{
    const char *archive = name2sccs(name, FALSE);

    (void) sp;
    (void) level;

    if (fexists(archive)) {
	if (debug)
	    track_wd(path);
	if (strcmp(path, working_path)) {
	    if (debug)
		PRINTF("\n");
	    SetWd(working_path);
	}
	checkout(name);
	if (strcmp(path, working_path))
	    SetWd(path);
    }

    return (readable);
}

/*ARGSUSED*/
static int
WALK_FUNC(scan_tree)
{
    char tmp1[MAXPATHLEN];
    char vault[MAXPATHLEN];
    char *s = pathcat(tmp1, path, name);
    Stat_t sb;

    (void) level;

    working_path = path;
    if (sp == 0 || readable < 0) {
	if (!dir_only) {
	    if (!fexists(name2sccs(name, FALSE))) {
		readable = -1;
		perror(name);
		if (!force)
		    exit(FAIL);
	    }
	    if (debug)
		track_wd(path);
	    checkout(name);
	}
    } else if (isDIR(sp->st_mode)) {
	abspath(s);		/* get rid of "." and ".." names */
	(void) strcpy(vault, sccs_dir(path, "."));
	if (!a_opt && *pathleaf(s) == '.') {
	    readable = -1;
	} else if (sameleaf(s, sccs_dir((char *) 0, (char *) 0))) {
	    /*
	     * patch: this assumes that even SCCS_VAULT will have
	     * leafnames that can be identified.
	     */
	    readable = -1;	/* require special scanning */
	    if (dir_only && (stat_dir(vault, &sb) >= 0)) {
		if (debug)
		    track_wd(vault);
		(void) walktree(path, vault, do_archive, "r", 0);
	    }
	} else if (sameleaf(s, rcs_dir(NULL, NULL))) {
	    readable = -1;
	}
	if (readable >= 0) {
	    if (debug)
		track_wd(path);
	}
    } else if (isFILE(sp->st_mode)) {
	if (!dir_only) {
	    if (debug)
		track_wd(path);
	    checkout(name);
	}
    } else {
	readable = -1;
    }

    return (readable);
}

static void
do_arg(const char *name)
{
    Stat_t sb;
    dir_only = (stat_dir(name, &sb) >= 0)
	|| sameleaf(sccs_dir((char *) 0, (char *) 0), name);
    (void) walktree((char *) 0, name, scan_tree, "r", 0);
}

static void
usage(void)
{
    static const char *tbl[] =
    {
	"Usage: sccsget [options] files_or_directories"
	,""
	,"Options"
	,"  -a       process all directories, including those beginning with \".\""
	,"  -n       traverse tree only, don't really extract (repeat to see lower"
	,"           level trace from getdelta)"
	,""
	,"Other options are passed to getdelta:"
	,"  -c DATE  specify cutoff date"
	,"  -e       extract for editing (suppress keywords, passed to \"get\")"
	,"  -k       suppress keyword expansion"
	,"  -r SID   specify revision"
    };
    unsigned j;
    for (j = 0; j < sizeof(tbl) / sizeof(tbl[0]); j++)
	FPRINTF(stderr, "%s\n", tbl[j]);
    FFLUSH(stderr);		/* for CLIX */
    exit(FAIL);
}

/*ARGSUSED*/
_MAIN
{
    int j;

    debug = sccs_debug();
    if (debug)
	track_wd((char *) 0);

    get_opts = argv_init();
    while ((j = getopt(argc, argv, "ac:efknr:s")) != EOF) {
	char tmp[BUFSIZ];
	FORMAT(tmp, "-%c", j);
#define	OPT	strcat(tmp, optarg)
	switch (j) {
	case 'a':
	    a_opt = TRUE;
	    break;
	case 'c':
	    argv_append(&get_opts, OPT);
	    break;
	case 'f':
	    force = TRUE;
	    /*FALLTHRU */
	case 'e':
	    /*FALLTHRU */
	case 'k':
	    argv_append(&get_opts, tmp);
	    break;
	case 'n':
	    no_op++;
	    break;
	case 'r':
	    argv_append(&get_opts, OPT);
	    break;
	default:
	    usage();
	}
    }
    if (!debug)
	argv_append(&get_opts, "-s");

    if (optind < argc) {
	for (j = optind; j < argc; j++)
	    do_arg(argv[j]);
    } else {
	do_arg(".");
    }

    exit(SUCCESS);
    /*NOTREACHED */
}
